Update SampleAggregator to use Accumlators

cmcantalupo opened the pull request.

cmcantalupo committed:
First draft of updated SampleAggregator

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Get Reporter to compile

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dianarg reviewed (commented):
Leaving a comment for when we come back to this.  Where does the knowledge about which signals are summed and which are averaged come from?  We talked about maybe having a new IOGroup API.  The existing agg_function is about aggregating domains, this is about how we agg over time.

cmcantalupo commented on test/SampleAggregatorTest.cpp:
```diff
@@ -250,31 +249,29 @@ TEST_F(SampleAggregatorTest, sample_total)
+        EXPECT_EQ(exp_time[0][region], m_agg->sample_region(M_SIGNAL_TIME, region));
+        EXPECT_EQ(exp_energy[0][region], m_agg->sample_region(M_SIGNAL_ENERGY_0, region));
+        EXPECT_EQ(exp_energy[1][region], m_agg->sample_region(M_SIGNAL_ENERGY_1, region));
+        EXPECT_EQ(exp_cycles[0][region], m_agg->sample_region(M_SIGNAL_CYCLES_0, region)) << region;
```
either make the region print verbose and for all errors, or delete.

dianarg commented on src/SampleAggregator.cpp:
```diff
@@ -61,108 +58,320 @@ namespace geopm
+                    signal.epoch_accum->enter();
+                }
+                // If we have observed our first epoch, update epoch totals
+                if (signal.epoch_count_last != 0) {
```
doesn't this skip updating for the first sample in the first epoch, since epoch is not updated until 206?

cmcantalupo replied:
Same logic as for region entry/exit

dianarg reviewed (commented):
Unit tests for averages would be nice; maybe we can come back to it.

cmcantalupo committed:
Make helper emplace function a template

- Similar helper function is a template strategy can be applied to the update helpers.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add some template helper functions for common code

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Clean up error messages in test

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>
